File: src/pattern/composite/java/Directory.java
package pattern.composite.java;

import java.util.ArrayList;

public class Directory extends Entry {
    private String name;

    private ArrayList<Entry> directory = new ArrayList<Entry>();

    public Directory(String name){
        this.name = name;
    }

    @Override
    public String getName() {
        return this.name;
    }

    @Override
    int getSize() {
        int totalSize  = 0;
        for (Entry it :
                this.directory) {
            totalSize += it.getSize();
        }

        return totalSize;
    }

    public Entry add(Entry entry){
        this.directory.add(entry);
        return entry;
    }

    @Override
    protected void printList(String prefix) {
        System.out.println(prefix+"/"+this);

        for (Entry it: this.directory){
            it.printList(prefix+"/"+this.name);
        }
    }
}

File: src/pattern/composite/kotlin/Directory.kt
package pattern.composite.kotlin

class Directory(var _name: String) : Entry(){
    val directory: ArrayList<Entry> = ArrayList<Entry>()

    override fun getName(): String {
        return this._name
    }

    override fun getSize(): Int {
        var totalSize = 0
        directory.map { totalSize += it.getSize() }
        return totalSize
    }

    override fun add(entry: Entry): Entry {
        this.directory.add(entry)
        return entry
    }

    override fun printList(prefix: String) {
        println("${prefix}/${this}")

        directory.forEach{ it.printList("${prefix}/${this._name}")}
    }
}